iceoryx_utils/platform/mac/source/semaphore.cpp

```diff
@@ -0,0 +1,248 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_utils/platform/semaphore.hpp"
+#include "iceoryx_utils/platform/time.hpp"
+
+#include <chrono>
+#include <cstdarg>
+#include <cstdlib>
+#include <iostream>
+#include <thread>
+
+static void deleteValue(std::atomic<int>*& value) noexcept
+{
+    if (value != nullptr)
+    {
+        delete value;
+        value = nullptr;
+    }
+}
+
+iox_sem_t::iox_sem_t() noexcept
+    : m_value{new std::atomic<int>}
+{
+}
+
+iox_sem_t::iox_sem_t(iox_sem_t&& rhs) noexcept
+{
+    *this = std::move(rhs);
+}
+
+iox_sem_t::~iox_sem_t()
+{
+    deleteValue(m_value);
+}
+
+iox_sem_t& iox_sem_t::operator=(iox_sem_t&& rhs) noexcept
+{
+    if (this != &rhs)
+    {
+        deleteValue(m_value);
+        m_handle = rhs.m_handle;
+        m_hasPosixHandle = rhs.m_hasPosixHandle;
+        m_value = rhs.m_value;
+        rhs.m_value = nullptr;
+    }
+    return *this;
+}
+
+int iox_sem_getvalue(iox_sem_t* sem, int* sval)
+{
+    *sval = sem->m_value->load();
+    return 0;
+}
+
+int iox_sem_post(iox_sem_t* sem)
+{
+    int retVal{0};
+    if (sem->m_hasPosixHandle)
+    {
+        retVal = sem_post(sem->m_handle.posix);
+    }
+    else
+    {
+        // dispatch semaphore always succeed
+        dispatch_semaphore_signal(sem->m_handle.dispatch);
+    }
+    sem->m_value->fetch_add((retVal == 0) ? 1 : 0);
+    return retVal;
+}
+
+int iox_sem_wait(iox_sem_t* sem)
+{
+    int retVal{0};
+    if (sem->m_hasPosixHandle)
+    {
+        retVal = sem_wait(sem->m_handle.posix);
+    }
+    else
+    {
+        // dispatch semaphore always succeed
+        dispatch_semaphore_wait(sem->m_handle.dispatch, DISPATCH_TIME_FOREVER);
+    }
+
+    sem->m_value->fetch_sub((retVal == 0) ? 1 : 0);
+    return retVal;
+}
+
+int iox_sem_trywait(iox_sem_t* sem)
+{
+    int retVal{0};
+    if (sem->m_hasPosixHandle)
+    {
+        retVal = sem_trywait(sem->m_handle.posix);
+    }
+    else
+    {
+        if (dispatch_semaphore_wait(sem->m_handle.dispatch, 0) != 0)
+        {
+            errno = EAGAIN;
+            retVal = -1;
+        }
+    }
+
+    sem->m_value->fetch_sub((retVal == 0) ? 1 : 0);
+    return retVal;
+}
+
+int iox_sem_timedwait(iox_sem_t* sem, const struct timespec* abs_timeout)
+{
+    struct timeval tv;
+    gettimeofday(&tv, nullptr);
+
+    int64_t timeoutInNanoSeconds =
+        (abs_timeout->tv_sec < tv.tv_sec
+         || (abs_timeout->tv_sec == tv.tv_sec && abs_timeout->tv_nsec < tv.tv_usec * NSEC_PER_USEC))
+            ? 0
+            : (abs_timeout->tv_sec - tv.tv_sec) * NSEC_PER_SEC + abs_timeout->tv_nsec - tv.tv_usec * NSEC_PER_USEC;
```

Comment:
How is the evaluation ordering? What happens if `tv_nsecs` is smaller than `tv_usec * NSEC_PER_USEC`?

Reply:
Evaluation order is from left to right, like the following:
1. abs_timeout->tv_sec < tv.tv_sec ||
2.1. abs_timeout->tv_sec == tv.tv_sec &&
2.2 abs_timeout->tv_nsec < tv.tv_usec * NSEC_PER_USEC

I see here no problem. If tv_nsecs is smaller than tv_usec than it is only substracted when abs_timeout->tv_sec is greater than tv.tv_sec.

But this little algorithm is too complicated. I replaced it with std::max(0, ...)